### DISCUSSION TITLE
Can't pass s_priceFeed as argument to getConversionRate() - Lesson 7 FUndMe

### DISCUSSION THREAD

**Discussion by Zubairuzb** [Cyfrin/foundry-full-course-cu#1948] [General]
Immediately i start refactoring by following along with the tutorial i got error on this function
```
   
    function fund() public payable {
        require(msg.value.getConversionRate(s_priceFeed) >= MINIMUM_USD, "You need to spend more ETH!");
        // require(PriceConverter.getConversionRate(msg.value) >= MINIMUM_USD, "You need to spend more ETH!");
        addressToAmountFunded[msg.sender] += msg.value;
        funders.push(msg.sender);
    }
```
## This is the error i am getting      
```
Invalid type for argument in function call. Invalid implicit conversion from contract AggregatorV3Interface to contract AggregatorV3Interface requested.solidity(9553)
AggregatorV3Interface private s_priceFeed
```

### Here is my FundMe contract 
```
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

// Note: The AggregatorV3Interface might be at a different location than what was in the video!
import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/shared/interfaces/AggregatorV3Interface.sol";
import {PriceConverter} from "./PriceConverter.sol";

error FundMe__NotOwner();

contract FundMe {
    using PriceConverter for uint256;

    mapping(address => uint256) public addressToAmountFunded;
    address[] public funders;

    // Could we make this constant?  /* hint: no! We should make it immutable! */
    address public /* immutable */ i_owner;
    uint256 public constant MINIMUM_USD = 5 * 10 ** 18;
    AggregatorV3Interface private s_priceFeed;
    
    constructor(address priceFeed) {
        i_owner = msg.sender;
        s_priceFeed = AggregatorV3Interface(priceFeed);
    }

    function fund() public payable {
        require(msg.value.getConversionRate(s_priceFeed) >= MINIMUM_USD, "You need to spend more ETH!");
        // require(PriceConverter.getConversionRate(msg.value) >= MINIMUM_USD, "You need to spend more ETH!");
        addressToAmountFunded[msg.sender] += msg.value;
        funders.push(msg.sender);
    }

    function getVersion() public view returns (uint256) {
        return s_priceFeed.version();
    }

    modifier onlyOwner() {
        // require(msg.sender == owner);
        if (msg.sender != i_owner) revert FundMe__NotOwner();
        _;
    }

    function withdraw() public onlyOwner {
        for (uint256 funderIndex = 0; funderIndex < funders.length; funderIndex++) {
            address funder = funders[funderIndex];
            addressToAmountFunded[funder] = 0;
        }
        funders = new address[](0);
        // // transfer
        // payable(msg.sender).transfer(address(this).balance);

        // // send
        // bool sendSuccess = payable(msg.sender).send(address(this).balance);
        // require(sendSuccess, "Send failed");

        // call
        (bool callSuccess,) = payable(msg.sender).call{value: address(this).balance}("");
        require(callSuccess, "Call failed");
    }
    // Explainer from: https://solidity-by-example.org/fallback/
    // Ether is sent to contract
    //      is msg.data empty?
    //          /   \
    //         yes  no
    //         /     \
    //    receive()?  fallback()
    //     /   \
    //   yes   no
    //  /        \
    //receive()  fallback()

    fallback() external payable {
        fund();
    }

    receive() external payable {
        fund();
    }
}

// Concepts we didn't cover yet (will cover in later sections)
// 1. Enum
// 2. Events
// 3. Try / Catch
// 4. Function Selector
// 5. abi.encode / decode
// 6. Hash with keccak256
// 7. Yul / Assembly
```
### This is the PriceConverter contract 
```
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";

// Why is this a library and not abstract?
// Why not an interface?
library PriceConverter {
    // We could make this public, but then we'd have to deploy it
    function getPrice(
        AggregatorV3Interface priceFeed
    ) internal view returns (uint256) {
        // Sepolia ETH / USD Address
        // https://docs.chain.link/data-feeds/price-feeds/addresses

        (, int256 answer, , , ) = priceFeed.latestRoundData();
        // ETH/USD rate in 18 digit
        return uint256(answer * 10000000000);
    }

    // 1000000000
    function getConversionRate(
        uint256 ethAmount,
        AggregatorV3Interface priceFeed
    ) internal view returns (uint256) {
        uint256 ethPrice = getPrice(priceFeed);
        uint256 ethAmountInUsd = (ethPrice * ethAmount) / 1000000000000000000;
        // the actual ETH/USD conversion rate, after adjusting the extra 0s.
        return ethAmountInUsd;
    }
}

```

**Comment by EngrPips** [comment]
Awesome, I am glad you unblocked yourself.

**Reply by Zubairuzb** [reply]
I see the issue still persists 😊 Don't know what to do

**Reply by EngrPips** [reply]
What were you trying to do before you ran into that error? Were you trying to compile or run a test? Your code seem fine to me.

**Reply by Zubairuzb** [reply]
It's just indicating the error in the code, and even if i am trying to compile it shows the same thng

**Reply by EngrPips** [reply]
It seems like you figured this out already

**Reply by Zubairuzb** [reply]
No

**Reply by EngrPips** [reply]
Alright, I will share with you my code and you can compare it with yours to see where you are making mistake.

**FundMe Contract**

```js
// SPDX-License-Identifier: UNLICENSED
pragma solidity 0.8.23;

import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";
import {PriceConverter} from "./PriceConverter.sol";
contract FundMe {

    using PriceConverter for uint256;

    uint256 constant MINIMUM_ALLOWED_VALUE = 5e18;
    address immutable  PRICE_FEED_ADDRESS;
    address payable immutable  i_owner; 

    address[] private s_funders;
    mapping(address funder => uint256 amount) private s_addressToAmountFunded;

    constructor(address  _priceFeedAddress) {
        PRICE_FEED_ADDRESS = _priceFeedAddress;
        i_owner = payable(msg.sender);
    }

    function fund() public payable {
        uint256 receivedValueInUSD = msg.value.getConversionRate(PRICE_FEED_ADDRESS);
        require(receivedValueInUSD >= MINIMUM_ALLOWED_VALUE, "You didn't send enough ether");
        s_funders.push(msg.sender);
        s_addressToAmountFunded[msg.sender] += msg.value; 
    }

    function withdraw() public {
        require(msg.sender == i_owner, "You are not the owner of the contract");
        uint256 fundersLength = s_funders.length;
        for(uint256 index; index < fundersLength; index++){
            address currentAddress = s_funders[index];
            s_addressToAmountFunded[currentAddress] = 0;
        }
        s_funders = new address[](0);
        (bool callSuccess, ) = i_owner.call{value: address(this).balance}("");
        require(callSuccess == true, "Sending ether to owner failed");
    }

    function getOwner() external view returns(address) {
        return i_owner;
    }

    function getPriceFeedVersion() external view returns(uint256) {
        return AggregatorV3Interface(PRICE_FEED_ADDRESS).version();
    }

    function getAddressToAmountFunded(address _address) external view returns(uint256) {
        return s_addressToAmountFunded[_address];
    }

    function getAddressAtIndex(uint256 _index) external view returns(address) {
        return s_funders[_index];
    }

    receive() external payable  {
        fund();
    }

    fallback() external payable { 
        fund();
    }
}

```

**PriceConverter Library**

```js
// SPDX-License-Identifier: UNLICENSED
pragma solidity 0.8.23;

import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";

library PriceConverter {
    // ETHUSD PRICEFEED ON SEPOLIA => 0x694AA1769357215DE4FAC081bf1f309aDC325306

    function getConversionRate(uint256 _ethValueToConvert, address _priceFeedAddress) internal view returns(uint256) {
        uint256 price = getPrice(_priceFeedAddress);
        uint256 receivedValueInUSD = _ethValueToConvert * (price * 10 ** 10);
        return receivedValueInUSD;
    }

    function getPrice(address _priceFeedAddress) internal view returns(uint256) {
        AggregatorV3Interface priceFeedContract = AggregatorV3Interface(_priceFeedAddress);
        (,int answer,,,) = priceFeedContract.latestRoundData();
        uint256 price = uint256(answer);
        return price;
    }

    
}
```

**Reply by EngrPips** [reply]
> Alright, I will share with you my code and you can compare it with yours to see where you are making mistake.
> 
> **FundMe Contract**
> 
> ```js
> // SPDX-License-Identifier: UNLICENSED
> pragma solidity 0.8.23;
> 
> import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";
> import {PriceConverter} from "./PriceConverter.sol";
> contract FundMe {
> 
>     using PriceConverter for uint256;
> 
>     uint256 constant MINIMUM_ALLOWED_VALUE = 5e18;
>     address immutable  PRICE_FEED_ADDRESS;
>     address payable immutable  i_owner; 
> 
>     address[] private s_funders;
>     mapping(address funder => uint256 amount) private s_addressToAmountFunded;
> 
>     constructor(address  _priceFeedAddress) {
>         PRICE_FEED_ADDRESS = _priceFeedAddress;
>         i_owner = payable(msg.sender);
>     }
> 
>     function fund() public payable {
>         uint256 receivedValueInUSD = msg.value.getConversionRate(PRICE_FEED_ADDRESS);
>         require(receivedValueInUSD >= MINIMUM_ALLOWED_VALUE, "You didn't send enough ether");
>         s_funders.push(msg.sender);
>         s_addressToAmountFunded[msg.sender] += msg.value; 
>     }
> 
>     function withdraw() public {
>         require(msg.sender == i_owner, "You are not the owner of the contract");
>         uint256 fundersLength = s_funders.length;
>         for(uint256 index; index < fundersLength; index++){
>             address currentAddress = s_funders[index];
>             s_addressToAmountFunded[currentAddress] = 0;
>         }
>         s_funders = new address[](0);
>         (bool callSuccess, ) = i_owner.call{value: address(this).balance}("");
>         require(callSuccess == true, "Sending ether to owner failed");
>     }
> 
>     function getOwner() external view returns(address) {
>         return i_owner;
>     }
> 
>     function getPriceFeedVersion() external view returns(uint256) {
>         return AggregatorV3Interface(PRICE_FEED_ADDRESS).version();
>     }
> 
>     function getAddressToAmountFunded(address _address) external view returns(uint256) {
>         return s_addressToAmountFunded[_address];
>     }
> 
>     function getAddressAtIndex(uint256 _index) external view returns(address) {
>         return s_funders[_index];
>     }
> 
>     receive() external payable  {
>         fund();
>     }
> 
>     fallback() external payable { 
>         fund();
>     }
> }
> ```
> 
> **PriceConverter Library**
> 
> ```js
> // SPDX-License-Identifier: UNLICENSED
> pragma solidity 0.8.23;
> 
> import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";
> 
> library PriceConverter {
>     // ETHUSD PRICEFEED ON SEPOLIA => 0x694AA1769357215DE4FAC081bf1f309aDC325306
> 
>     function getConversionRate(uint256 _ethValueToConvert, address _priceFeedAddress) internal view returns(uint256) {
>         uint256 price = getPrice(_priceFeedAddress);
>         uint256 receivedValueInUSD = _ethValueToConvert * (price * 10 ** 10);
>         return receivedValueInUSD;
>     }
> 
>     function getPrice(address _priceFeedAddress) internal view returns(uint256) {
>         AggregatorV3Interface priceFeedContract = AggregatorV3Interface(_priceFeedAddress);
>         (,int answer,,,) = priceFeedContract.latestRoundData();
>         uint256 price = uint256(answer);
>         return price;
>     }
> 
>     
> }
> ```

@TSEmersonCode, Please take a look at this code and compare it with yours to see where you get it wrong.

**Reply by vigneshbalu96** [reply]
It works but I really wanted to know why and How coz when I google/brave for the AggregatorV3Interface.sol it takes me to this link "https://github.com/smartcontractkit/chainlink/blob/develop/contracts/src/v0.8/shared/interfaces/AggregatorV3Interface.sol" but when I searched for the path "chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol" in the same repository `AggregatorV3Interface` really doesn't exist, could you show some light on this? @EngrPips

**Reply by EngrPips** [reply]
Hello @vigneshbalu96, You know codebase gets updated from time to time, so maybe the location of the file in the repository has changed or something.

**Comment by TSEmersonCode** [comment]
Hi !! Please did you find a solution to that? i have the same problem

**Comment by Zubairuzb** [comment]
Yes, i have resolve the issue, the problem is the AggregatorV3Interface import path for FundMe and PriceConverter are not the same

**Reply by EngrPips** [reply]
Glad you figured it out. Keep crushing it.

**Reply by Zubairuzb** [reply]
Thank you, that is lesson 7 seems difficult, any advice ?

**Reply by EngrPips** [reply]
Take your time. Please don't rush things at all. It takes time for our brain to absorb and cement whatever we learn, and with time, it becomes easier.

**Reply by Zubairuzb** [reply]
Alright. Thanks a lot, i will follow it gradually

**Reply by EngrPips** [reply]
You are welcome. I have been doing the `Merkle Tree Airdrop` course on cyfrin for about 5 days now, and it's just around 3 hours or so, I think. It's taking me that long because I am taking time to think, reason, and understand it properly. I learned to be patient the hard way, lol.

**Reply by Zubairuzb** [reply]
Great. I also follow similar pattern with yours to understand things better

**Reply by EngrPips** [reply]
Good things take time. Absorbing deep and sound knowledge takes even more time, lol
